File: FbxOpenCpp/GeometryTest.cpp

#include "Tests.h"

using namespace std;

void Geometry_AddDeformer_AddsDeformer()
{
    // given:
    FbxManager* manager = FbxManager::Create();
    FbxGeometry* g = FbxGeometry::Create(manager, "");
    FbxSkin* skin = FbxSkin::Create(manager, "");

    // require:
    AssertEqual(0, g->GetDeformerCount());
    AssertEqual(0, g->GetSrcObjectCount());

    // when:
    g->AddDeformer(skin);

    // then:
    AssertEqual(1, g->GetDeformerCount());
    AssertEqual(skin, g->GetDeformer(0));
    AssertEqual(1, g->GetSrcObjectCount());
    AssertEqual(skin, g->GetSrcObject(0));
}

void Geometry_AddSrcConnection_AddsDeformer()
{
    // given:
    FbxManager* manager = FbxManager::Create();
    FbxGeometry* g = FbxGeometry::Create(manager, "");
    FbxSkin* skin = FbxSkin::Create(manager, "");

    // require:
    AssertEqual(0, g->GetDeformerCount());
    AssertEqual(0, g->GetSrcObjectCount());

    // when:
    g->ConnectSrcObject(skin);

    // then:
    AssertEqual(1, g->GetDeformerCount());
    AssertEqual(skin, g->GetDeformer(0));
    AssertEqual(1, g->GetSrcObjectCount());
    AssertEqual(skin, g->GetSrcObject(0));
}

void Geometry_RemoveDeformer_RemovesDeformer()
{
    // given:
    FbxManager* manager = FbxManager::Create();
    FbxGeometry* g = FbxGeometry::Create(manager, "");
    FbxSkin* skin = FbxSkin::Create(manager, "");
    g->AddDeformer(skin);

    // require:
    AssertEqual(1, g->GetDeformerCount());
    AssertEqual(1, g->GetSrcObjectCount());

    // when:
    g->RemoveDeformer(0);

    // then:
    AssertEqual(0, g->GetDeformerCount());
    AssertEqual(0, g->GetSrcObjectCount());
}

void Geometry_DisconnectSrcObject_RemovesDeformer()
{
    // given:
    FbxManager* manager = FbxManager::Create();
    FbxGeometry* g = FbxGeometry::Create(manager, "");
    FbxSkin* skin = FbxSkin::Create(manager, "");
    g->AddDeformer(skin);

    // require:
    AssertEqual(1, g->GetDeformerCount());
    AssertEqual(1, g->GetSrcObjectCount());

    // when:
    g->DisconnectSrcObject(skin);

    // then:
    AssertEqual(0, g->GetDeformerCount());
    AssertEqual(0, g->GetSrcObjectCount());
}

void GeometryTest::RegisterTestCases()
{
    AddTestCase(Geometry_AddDeformer_AddsDeformer);
    AddTestCase(Geometry_AddSrcConnection_AddsDeformer);
    AddTestCase(Geometry_RemoveDeformer_RemovesDeformer);
    AddTestCase(Geometry_DisconnectSrcObject_RemovesDeformer);
}

File: FbxSharpTests/GeometryTest.cs
using System;
using NUnit.Framework;
using FbxSharp;

namespace FbxSharpTests
{
    [TestFixture]
    public class GeometryTest : TestBase
    {
        [Test]
        public void Geometry_AddDeformer_AddsDeformer()
        {
            // given:
            var g = new FbxMesh("");
            var skin = new FbxSkin("");

            // require:
            Assert.AreEqual(0, g.GetDeformerCount());
            Assert.AreEqual(0, g.GetSrcObjectCount());

            // when:
            g.AddDeformer(skin);

            // then:
            Assert.AreEqual(1, g.GetDeformerCount());
            Assert.AreSame(skin, g.GetDeformer(0));
            Assert.AreEqual(1, g.GetSrcObjectCount());
            Assert.AreSame(skin, g.GetSrcObject(0));
        }

        [Test]
        public void Geometry_AddSrcConnection_AddsDeformer()
        {
            // given:
            var g = new FbxMesh("");
            var skin = new FbxSkin("");

            // require:
            Assert.AreEqual(0, g.GetDeformerCount());
            Assert.AreEqual(0, g.GetSrcObjectCount());

            // when:
            g.ConnectSrcObject(skin);

            // then:
            Assert.AreEqual(1, g.GetDeformerCount());
            Assert.AreSame(skin, g.GetDeformer(0));
            Assert.AreEqual(1, g.GetSrcObjectCount());
            Assert.AreSame(skin, g.GetSrcObject(0));
        }

        [Test]
        public void Geometry_RemoveDeformer_RemovesDeformer()
        {
            // given:
            var g = new FbxMesh("");
            var skin = new FbxSkin("");
            g.AddDeformer(skin);

            // require:
            Assert.AreEqual(1, g.GetDeformerCount());
            Assert.AreEqual(1, g.GetSrcObjectCount());

            // when:
            g.RemoveDeformer(0);

            // then:
            Assert.AreEqual(0, g.GetDeformerCount());
            Assert.AreEqual(0, g.GetSrcObjectCount());
        }

        [Test]
        public void Geometry_DisconnectSrcObject_RemovesDeformer()
        {
            // given:
            var g = new FbxMesh("");
            var skin = new FbxSkin("");
            g.AddDeformer(skin);

            // require:
            Assert.AreEqual(1, g.GetDeformerCount());
            Assert.AreEqual(1, g.GetSrcObjectCount());

            // when:
            g.DisconnectSrcObject(skin);

            // then:
            Assert.AreEqual(0, g.GetDeformerCount());
            Assert.AreEqual(0, g.GetSrcObjectCount());
        }
    }
}
